src/oc_erchef/apps/chef_objects/src/chef_cookbook_version.erl

```diff
@@ -589,6 +672,84 @@ list_query(_ObjectRec) ->
 bulk_get_query(_ObjectRec) ->
     error(not_implemented).
 
+%% for v2 metadata, the "segment" is encoded in to the name of file
+%% so for data that is v0 on disk we have to prepend the segment to the name
+add_segment_to_filename(Segment, File) ->
+    Fn = ej:get({<<"name">>}, File),
+    Fn1 = iolist_to_binary([Segment, "/", Fn]),
+    ej:set({<<"name">>}, File, Fn1).
+
+remove_segment_from_filename(File) ->
+    [Segment | Name ] = get_segment_from_record(File),
+    case Name of
+        % if Name is nil, then we have a root file (like metadata.rb) - so we'll set the name to the segment, and the segment to <<"root_files">>
+        [] ->
+            { <<"root_files">>, ej:set({<<"name">>}, File, Segment)};
+        _ ->
+            Record = ej:set({<<"name">>}, File, lists:last(Name)),
+            { Segment, Record }
+    end.
+
+populate_all_files(Segment, Data, Metadata) ->
+    lists:foldl(fun(File, CB) ->
+                        File1 = case Segment of
+                                    <<"root_files">> ->
+                                        File;
+                                    _ ->
+                                        add_segment_to_filename(Segment, File)
+                                end,
+                        ej:set_p({<<"all_files">>, new}, CB, File1)
+                end,
+                Metadata,
+                Data).
+
+%% A v2 cookbook version contains only the "all_files" key, which is a list of all the file parts
+%% { "all_files": [ { "name": "recipes/default.rb", "path": "recipes/default.rb", … } ] }
+%% We have to transform that into segments (?COOKBOOK_SEGMENTS) containing a list of file parts for that segment
+%% { "recipes": [ { "name": "default.rb", path: "recipes/default.rb", … } ] }
+populate_segments(Data, Metadata) ->
+    CB1 = lists:foldl(fun(Segment, CB) ->
+                              ej:set({Segment}, CB, [])
+                      end,
+                      Metadata,
+                      ?COOKBOOK_SEGMENTS),
+
+    lists:foldl(fun(File, CB) ->
+                        { Segment, Record } = remove_segment_from_filename(File),
+                        case lists:member(Segment, ?COOKBOOK_SEGMENTS) of
```

Comment:
What happens to files that are uploaded via a V2 call that don't fit in the V0 segment pattern? The RFC doesn't mention that, but I see @stevendanna commented about it (https://github.com/chef/chef-rfc/pull/161#issuecomment-156105423). What was the decision? It looks like we silently drop them; I wonder if we might want to error instead, or otherwise we could silently be breaking cookbooks.

Reply:
I think the only reasonable thing we can do is drop them; our choice is to drop them - and have cookbooks that only encompass the functionality that chef 12 can provide, whilst upgrading to 13 will handle everything; or we error, at which point cookbooks uploaded on 13 will almost always entirely stop working on 12, because inevitably we'll upload something (like a test directory) that 12 doesn't understand but that doesn't affect running the cookbook.

Reply:
Understand the reasoning, but it strikes me as creating a really painful debug experience where those files are needed in the cookbook, but silently dropped on an older version of chef. 
This is where something like a 'required_chef_version' cookbook metadata field would be nice; you could allow uploading non-standard paths only if the required_chef_version was set for a version that supported , and error on downloading cookbooks to chef versions older than supported.